### ISSUE TITLE
Enhancements for Live Migration - Cross-Host TCP Migration, Memory Optimization, and QoS Traffic Control

### ISSUE THREAD

**Issue by ljrcore** [cloud-hypervisor/cloud-hypervisor#6840]
I am currently working on enhancing the live migration capabilities in Cloud Hypervisor. The proposed improvements aim to reduce downtime, optimize memory migration, and implement QoS traffic control. The main contents includeinclude:

1. Cross-host TCP live migration
2. Memory migration optimization
3. QoS traffic control

Here is the current TODO list for enhancing the live migration capabilities in Cloud Hypervisor:
- [x] Implement ReadVolatile and WriteVolatile for TcpStream
- [x] Implement TCP-based cross-host live migration to minimize downtime and total migration time
- [ ] Enhance memory pre-copy and add downtime limit function
- [ ] Introduce Quality of Service (QoS) control to manage and limit migration traffic, ensuring minimal impact on network performance

I plan to start with the implementation of the cross-host TCP migration feature and will provide updates on progress and any preliminary results.

Feedback and suggestions are welcomed and appreciated.

Thanks

**Comment by liuw** [comment]
Can you share more details about 1? Are you going to make the source Cloud Hypervisor instance connect directly to the destination instance? I suppose that's going to avoid a lot of extra copying compared to using `nc` or `socat`.

Do you plan to add TLS support?

**Comment by rbradford** [comment]
Using only a unix socket for live migration is an intentional design choice since using a plain TCP socket is a security problem and to add SSL would be a huge complexity (certificate validation, certificate stores, etc) not to mention you might want some authentication and authorization on top. This should all be part of the management layer and so by using a unix socket it can proxy it howevers it sees fit.

We simply use socat in our examples for demonstration and testing purposes - I wouldn't expect anybody to use that in production.

**Comment by liuw** [comment]
> Using only a unix socket for live migration is an intentional design choice since using a plain TCP socket is a security problem and to add SSL would be a huge complexity (certificate validation, certificate stores, etc) not to mention you might want some authentication and authorization on top. This should all be part of the management layer and so by using a unix socket it can proxy it howevers it sees fit.

Having another tool to copy things around can introduces significant performance issues with large VMs. If both sides are trusted (no TLS required), one can still benefit from a TCP based migration scheme, right?

**Comment by rbradford** [comment]
> > Using only a unix socket for live migration is an intentional design choice since using a plain TCP socket is a security problem and to add SSL would be a huge complexity (certificate validation, certificate stores, etc) not to mention you might want some authentication and authorization on top. This should all be part of the management layer and so by using a unix socket it can proxy it howevers it sees fit.
> 
> Having another tool to copy things around can introduces significant performance issues with large VMs. If both sides are trusted (no TLS required), one can still benefit from a TCP based migration scheme, right?

I don't see why you would ever transfer the contents of a VM's memory unencrypted even over a network you think is trusted - the unix socket solution allows the use of encryption and authentication.

It's only the first copy that is particularly bulky (and again using a smart proxy would allow you to compress the zeroes for the unused pages if you so desired.)

**Comment by ljrcore** [comment]
> Can you share more details about 1? Are you going to make the source Cloud Hypervisor instance connect directly to the destination instance? I suppose that's going to avoid a lot of extra copying compared to using `nc` or `socat`.

Yes, the source host can be directly migrated to the destination host via tcp, without `nc` or `socat`.

> 
> Do you plan to add TLS support?

The basic version I am developing does not support TLS. Whether it will be supported in the future needs further discussion.

**Comment by ljrcore** [comment]
> Using only a unix socket for live migration is an intentional design choice since using a plain TCP socket is a security problem and to add SSL would be a huge complexity (certificate validation, certificate stores, etc) not to mention you might want some authentication and authorization on top. This should all be part of the management layer and so by using a unix socket it can proxy it howevers it sees fit.
> 
> We simply use socat in our examples for demonstration and testing purposes - I wouldn't expect anybody to use that in production.

We can support both tcp and unix for cross-host live migration. Users can choose tcp or unix during live migration. I will contribute a basic version that supports TCP, which does not affect the original Unix function.

From the test results, migration performance has been greatly improved, especially under memory pressure. I still need some time to optimize the code, develop integration tests, and perform performance tests. I will contribute it after completion.

**Comment by up2wing** [comment]
Would you like to organize the things what you want to do as TODO list? So people can get the whole progress of the issue more easily. 
Thx.

**Comment by up2wing** [comment]
> > Using only a unix socket for live migration is an intentional design choice since using a plain TCP socket is a security problem and to add SSL would be a huge complexity (certificate validation, certificate stores, etc) not to mention you might want some authentication and authorization on top. This should all be part of the management layer and so by using a unix socket it can proxy it howevers it sees fit.
> 
> Having another tool to copy things around can introduces significant performance issues with large VMs. If both sides are trusted (no TLS required), one can still benefit from a TCP based migration scheme, right?

Agree.

As far as I know, many public cloud or private cloud vendors usually migrate VMs over their internal networks, and security is guaranteed by the infrastructure.  We can provide this new way, allowing users to choose which one to use, and when they make a choice, they should be aware of the potential risks.

**Comment by ljrcore** [comment]
> Would you like to organize the things what you want to do as TODO list? So people can get the whole progress of the issue more easily. Thx.

Thanks for your suggestion. I think it is a good idea to organize tasks into TODO lists. I will update the first message in this issue to add the TODO list.

**Comment by ljrcore** [comment]
The latest list is here:

https://github.com/cloud-hypervisor/cloud-hypervisor/issues/7111
